src/lib/user_command_input/user_command_input.ml

```diff
@@ -0,0 +1,206 @@
+  | Keypair sender_kp ->
+      Deferred.Result.return (User_command.sign sender_kp user_command_payload)
+  | Hd_index hd_index ->
+      Secrets.Hardware_wallets.sign ~hd_index
+        ~public_key:(Public_key.decompress_exn sender)
+        ~user_command_payload
+
+let inferred_nonce ~f ~(sender : Public_key.Compressed.t) ~nonce_map =
+  let update_map = Public_key.Compressed.Map.set nonce_map ~key:sender in
+  match Public_key.Compressed.Map.find nonce_map sender with
+  | Some nonce ->
+      (*multiple user commands from a sender*)
+      let next_nonce = Account_nonce.succ nonce in
+      let updated_map = update_map ~data:next_nonce in
+      Ok (next_nonce, updated_map)
+  | None -> (
+    match Participating_state.active (f sender) |> Option.join with
+    | None ->
+        Error
+          "Couldn't infer nonce for transaction from specified `sender` since \
+           `sender` is not in the ledger or sent a transaction in transaction \
+           pool."
+    | Some txn_pool_or_account_nonce ->
+        let updated_map = update_map ~data:txn_pool_or_account_nonce in
+        Ok (txn_pool_or_account_nonce, updated_map) )
+
+let to_user_command ~result_cb ~infer_nonce ~logger uc_inputs :
+    (User_command.t list * (_ Or_error.t -> unit)) option Deferred.t =
+  let setup_user_command (client_input : t) nonce_map =
+    (*create a user_command from input*)
+    let open Deferred.Or_error.Let_syntax in
+    let sender = client_input.sender in
+    let to_or_error : type a. (a, string) Result.t -> a Or_error.t =
+     fun e ->
+      let err_with_input str =
+        sprintf "Error creating user command: %s Error: %s"
+          (Yojson.Safe.to_string (to_yojson client_input))
+          str
+      in
+      Result.map_error e ~f:(Fn.compose Error.of_string err_with_input)
+    in
+    let%bind inferred_nonce, updated_nonce_map =
+      inferred_nonce ~f:infer_nonce ~sender ~nonce_map
+      |> to_or_error |> Deferred.return
+    in
+    let%bind user_command_payload =
+      Payload.to_user_command_payload client_input.payload ~inferred_nonce
+      |> to_or_error |> Deferred.return
+    in
+    let%map signed_user_command =
+      Deferred.map
+        (sign ~sender:client_input.sender ~user_command_payload
+           client_input.signature)
+        ~f:to_or_error
+    in
+    (User_command.forget_check signed_user_command, updated_nonce_map)
+  in
+  let%map user_commands =
+    (*When batching multiple user commands, keep track of the nonces and send all the user commands if they are valid or none if there is an error in one of them*)
+    Deferred.Or_error.List.fold ~init:([], Public_key.Compressed.Map.empty)
+      uc_inputs ~f:(fun (valid_user_commands, nonce_map) uc_input ->
+        let open Deferred.Or_error.Let_syntax in
+        let%map res, updated_nonce_map =
+          setup_user_command uc_input nonce_map
+        in
+        (res :: valid_user_commands, updated_nonce_map) )
+  in
+  match user_commands with
```

Comment:
This logic is specific to the callsite in `Coda_lib`, please move it there instead.